autofaiss/external/distributed.py

```diff
@@ -0,0 +1,367 @@
+
+    b = faiss.read_index(tmp_name)
+    os.remove(tmp_name)
+    return b
+
+
+def _get_bytes_from_index(index: faiss.Index) -> bytearray:
+    """Transforms a faiss index into a bytearray."""
+
+    with NamedTemporaryFile(delete=False) as output_file:
+        faiss.write_index(index, output_file.name)
+        tmp_name = output_file.name
+
+    with open(tmp_name, "rb") as index_file:
+        b = index_file.read()
+        os.remove(tmp_name)
+        return bytearray(b)
+
+
+def _generate_small_index_file_name(batch_id: int) -> str:
+    return f"index_{batch_id}"
+
+
+def _save_small_index(index: faiss.Index, batch_id: int, small_indices_folder: str):
+    """Save index for one batch."""
+    fs = _get_file_system(small_indices_folder)
+    fs.mkdirs(small_indices_folder, exist_ok=True)
+    small_index_filename = _generate_small_index_file_name(batch_id)
+    with fsspec.open(small_index_filename, "wb").open() as f:
+        faiss.write_index(index, faiss.PyCallbackIOWriter(f.write))
+    dest_filepath = os.path.join(small_indices_folder, small_index_filename)
+    fs.put(small_index_filename, dest_filepath)
+
+
+def _add_index(
+    start: int,
+    end: int,
+    broadcast_trained_index_bytes,
+    file_system: fsspec.AbstractFileSystem,
+    chunk_sizes: List[int],
+    embeddings_file_paths: List[str],
+    embedding_column_name: str,
+    batch_id: int,
+    small_indices_folder: str,
+    file_format: str,
+    id_columns: Optional[List[str]] = None,
+    num_cores: Optional[int] = None,
+    embedding_ids_df_handler: Optional[Callable[[pd.DataFrame, int], Any]] = None,
+):
+    """
+    Add a batch of embeddings on trained index and save this index.
+
+    Parameters
+    ----------
+    start: int
+        Start position of this batch
+    end: int
+        End position of this batch
+    broadcast_trained_index_bytes: pyspark.Broadcast
+        Trained yet empty index
+    chunk_sizes: List[int]
+        A list of number of vectors in each embedding files
+    embeddings_file_paths: List[str]
+        A list of embeddings file paths
+    embedding_column_name: str
+        Embeddings column name for parquet ; default "embedding"
+    batch_id: int
+        Batch id
+    small_indices_folder: str
+        The folder where we save all the small indices
+    num_cores: int
+        Number of CPU cores (not Vcores)
+    file_format: str
+        Embedding file format, "npy" or "parquet"
+    id_columns: Optional[List[str]]
+        Names of the columns containing the Ids of the vectors, only used when file_format is "parquet"
+    embedding_ids_df_handler: Optional[Callable[[pd.DataFrame, int], Any]]
+        The function that handles the embeddings Ids when id_columns is given
+    """
+    if end > sum(chunk_sizes):
+        end = sum(chunk_sizes)
+    if len(chunk_sizes) != len(embeddings_file_paths):
+        raise ValueError("The length of chunk_sizes should be equal to the number of embeddings_file_paths")
```

Comment:
Maybe you should output both len(chunk_sizes) and len(embeddings_file_paths) to help debugging

Reply:
Sounds good to me. They are actually what I have done in debugging.